src/ops/indexing.rs

```diff
@@ -0,0 +1,1799 @@
+
+                gather_indices.push(Rc::new(item));
+            }
+            TakeArray { indices } => {
+                is_slice.push(false);
+                max_dims = max_dims.max(indices.ndim());
+                gather_indices.push(Rc::clone(indices));
+            }
+            Ellipsis | ExpandDims => {
+                unreachable!("Unexpected operation in gather_nd")
+            }
+        }
+    }
+
+    // reshape them so that the int/array indices are first
+    if gather_first {
+        if slice_count > 0 {
+            let mut slice_index = 0;
+            for (i, item) in gather_indices.iter_mut().enumerate() {
+                if is_slice[i] {
+                    let mut new_shape = vec![1; max_dims + slice_count];
+                    new_shape[max_dims + slice_index] = item.dim(0);
+                    *item = Rc::new(item.reshape(&new_shape));
+                    slice_index += 1;
+                } else {
+                    let mut new_shape = item.shape().to_vec();
+                    new_shape.extend((0..slice_count).map(|_| 1));
+                    *item = Rc::new(item.reshape(&new_shape));
+                }
+            }
+        }
+    } else {
+        // reshape them so that the int/array indices are last
+        for (i, item) in gather_indices[..slice_count].iter_mut().enumerate() {
+            let mut new_shape = vec![1; max_dims + slice_count];
+            new_shape[i] = item.dim(0);
+            *item = Rc::new(item.reshape(&new_shape));
+        }
+    }
+
+    // Do the gather
+    // let indices = new_mlx_vector_array(gather_indices);
+    // SAFETY: indices will be freed at the end of this function. The lifetime of the items in
+    // `gather_indices` is managed by the `gather_indices` vector.
+    let indices = unsafe {
+        let c_vec_array = mlx_sys::mlx_vector_array_new();
+        for item in gather_indices.iter() {
+            mlx_sys::mlx_vector_array_add(c_vec_array, item.c_array);
+        }
+        c_vec_array
+    };
+    let mut slice_sizes = shape.to_vec();
+    (0..operation_len).for_each(|i| slice_sizes[i] = 1);
+
+    let gathered = unsafe {
+        let c_array = mlx_sys::mlx_gather(
+            src.c_array,
+            indices,
+            axes.as_ptr(),
+            axes.len(),
+            slice_sizes.as_ptr(),
+            slice_sizes.len(),
+            stream.as_ptr(),
+        );
+
+        Array::from_ptr(c_array)
+    };
+    let gathered_shape = gathered.shape();
+
+    // Squeeze the dims
+    let output_shape: Vec<i32> = gathered_shape[0..(max_dims + slice_count)]
+        .iter()
+        .chain(gathered_shape[(max_dims + slice_count + operation_len)..].iter())
+        .copied()
+        .collect();
+    let result = gathered.reshape(&output_shape);
+
+    unsafe {
+        mlx_sys::mlx_free(indices as *mut _);
+    }
+
+    (max_dims, result)
+}
+
+#[inline]
+fn get_item_int(src: &Array, index: i32, axis: i32, stream: StreamOrDevice) -> Array {
```

Comment:
probably should be named: `get_item_index`, right?
Also index, should be getting resolved.

Reply:
> Also index, should be getting resolved.

True. I guess it didn't throw because `mlx` allows out of bound indexing

Reply:
Addressed in 656e701aec731fb757b7ff1a8a20352d8518612f